### PR TITLE
Metrics reporter

### PR DESCRIPTION
**Pull request by jeffgrunewald**
Adds a metrics reporter library to be implemented in the application services and automatically creates a webserver exposing those metrics with prometheus-optimized naming conventions on the `/metrics` endpoint.

Extend the specific metrics exported by overriding the behaviour's default `metrics/0` function to list the metrics desired.

### CODE REVIEW

File: apps/metrics_reporter/mix.exs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  def application do
+    [
+      extra_applications: [:logger]
+    ]
+  end
+
+  defp deps do
+    [
+      {:telemetry_poller, "~> 0.5.0"},
+      {:telemetry_metrics_prometheus, github: "jeffgrunewald/telemetry_metrics_prometheus"},
```
[the remark this answers is not in the document]
Comment (jeffgrunewald) [reply apps/metrics_reporter/mix.exs:27]: This is a debugging step. The upstream library currently doesn’t support multiple supervisors in the same deployment.

File: apps/metrics_reporter/lib/metrics_reporter.ex [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  Defines a behaviour and default implementation of a
+  process for handling and reporting application metrics
+  gathered from telemetry events to a webserver for scraping
+  by a Prometheus instance.
+  """
+
+  @callback metrics() :: [
+              Telemetry.Metrics.Counter.t()
+              | Telemetry.Metrics.Distribution.t()
+              | Telemetry.Metrics.LastValue.t()
```
[the remark this answers is not in the document]
Comment (jeffgrunewald) [reply apps/metrics_reporter/lib/metrics_reporter.ex:12]: Correct; the default only returns the lastvalue but you’re expected to override and append to that list.

File: apps/platform_runner/test/end_to_end_test.exs [hunk 3]
```diff
@@ -479,4 +479,13 @@ defmodule PlatformRunner.EndToEndTest do
       end
     end
   end
+
+  describe "metrics" do
+    test "exposes metrics for all services" do
+      for port <- [9569, 9570, 9572, 9573, 9574, 9576] do
```
Comment (jeffgrunewald) [inline apps/platform_runner/test/end_to_end_test.exs:485]: i assigned 9571 and 9575 to `define` and `profile` respectively going through the list and i'm pretty sure they're not currently running during the e2e test. When they are, this list should just be converted to a range object (`9569..9576`)

File: helm/presets/local.yaml [hunk 4]
```diff
@@ -1,8 +1,8 @@
 receive:
+  service_udp:
+    type: LoadBalancer
```
Comment (jeffgrunewald) [inline helm/presets/local.yaml:3]: exposing a service locally using docker's built-in k8s exposes the service to the host machine without any extra networking shenanigans.

File: helm/templates/receive/deployment.yaml [hunk 5]
```diff
@@ -33,11 +33,14 @@ spec:
             - start
           imagePullPolicy: {{ .Values.image.pullPolicy }}
           ports:
-            - name: http
```
Comment (jeffgrunewald) [inline helm/templates/receive/deployment.yaml:36]: this seemed like poor naming/assignment as HTTP is over TCP and websockets are typically just run on standard 80/443. swapped them for clarity.

File: apps/metrics_reporter/README.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
+# MetricsReporter
```
Comment (ManApart) [inline apps/metrics_reporter/README.md:1]: Thank you for documenting.

File: apps/service_profile/config/test.exs [hunk 7]
```diff
@@ -20,3 +20,5 @@ config :service_profile, Profile.Application,
     ],
     dispatcher: Brook.Dispatcher.Noop
   ]
+
+config :service_persist, Persist.MetricsReporter, port: 9574
```
Comment (jmorris-pillar) [inline apps/service_profile/config/test.exs:24]: Why is this a different port for test and integration whileas broadcast uses the same port for both?

File: helm/presets/aws.yaml [hunk 8]
```diff
@@ -1,7 +1,7 @@
 receive:
   service:
     type: ClusterIP
-  service_tcp:
```
Comment (jmorris-pillar) [inline helm/presets/aws.yaml:4]: why this change?

File: helm/templates/receive/deployment.yaml (hunk 5 above)
Comment (jmorris-pillar) [reply helm/templates/receive/deployment.yaml:36]: makes sense to me

File: helm/presets/aws.yaml (hunk 8 above)
Comment (jmorris-pillar) [reply helm/presets/aws.yaml:4]: I understand now

File: apps/metrics_reporter/README.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+Create an instance of the MetricsReporter by
+including the `use MetricsReporter` macro in a
+module within an application and include the
+custom module in your application supervision tree.
+
+The MetricsReporter comes with Erlang VM metrics
+reported by default but you can customize the metrics
+exported by a given instance of MetricsReporter by
+overriding the `metrics/0` function in the module to
```
[the remark this answers is not in the document]
Comment (jeffgrunewald) [reply apps/metrics_reporter/README.md:17]: done; good call

File: apps/metrics_reporter/README.md (hunk 6 above)
Comment (jeffgrunewald) [reply apps/metrics_reporter/README.md:1]: 👍

File: apps/service_profile/config/test.exs (hunk 7 above)
Comment (jeffgrunewald) [reply apps/service_profile/config/test.exs:24]: b/c i screwed up. this port is in use by `persist`

Comment (jeffgrunewald) [reply apps/service_profile/config/test.exs:24]: good catch

File: helm/templates/receive/service-udp.yaml [hunk 10]
```diff
@@ -1,17 +1,17 @@
 apiVersion: v1
 kind: Service
 metadata:
-  name: {{ include "hindsight.fullname" . }}-receive-tcp
+  name: {{ include "hindsight.fullname" . }}-receive-udp
```
Comment (ManApart) [inline helm/templates/receive/service-udp.yaml:4]: Why the switch from tcp to udp here?

Comment (jeffgrunewald) [reply helm/templates/receive/service-udp.yaml:4]: see my comment on the `receive/deployment.yaml` file above.
